### PR TITLE
Update github.com/thanos-io/objstore

### PR DESCRIPTION
**Pull request by andyasp**
#### What this PR does

Updates objstore, which required code changes to bring forward. The motivation for the update is to bring in a [GCS listing optimization](https://www.github.com/thanos-io/objstore/pull/76).

A choice I made was to not expose the `storage_connection_string` option objstore added for Azure. The target use case for it seems to be testing with Azurite, so my gut feeling was that it wasn't currently necessary and it could always be added later.

#### Which issue(s) this PR fixes or relates to

Fixes: N/A

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storage/tsdb/block/global_markers_bucket_client.go [hunk 1]
```diff
@@ -131,9 +131,9 @@ func (b *globalMarkersBucket) IsObjNotFoundErr(err error) bool {
 	return b.parent.IsObjNotFoundErr(err)
 }
 
-// IsCustomerManagedKeyError implements objstore.Bucket.
-func (b *globalMarkersBucket) IsCustomerManagedKeyError(err error) bool {
-	return b.parent.IsCustomerManagedKeyError(err)
+// IsAccessDeniedError implements objstore.Bucket.
```
Comment (andyasp) [inline pkg/storage/tsdb/block/global_markers_bucket_client.go:134]:
```suggestion
// IsAccessDeniedErr implements objstore.Bucket.
```

**Comment by andyasp** [comment]
Test failure looks legitimate and there's a data race in the logs too. I'll take a closer look tomorrow.

**Comment by pracucci** [comment]
> A choice I made was to not expose the storage_connection_string option objstore added for Azure. The target use case for it seems to be testing with Azurite, so my gut feeling was that it wasn't currently necessary and it could always be added later.

LGTM. As you already said, we can always add it later if needed.

**Comment by andyasp** [comment]
This looks like an issue with upstream. This diff allows the test to pass:
```diff
diff --git a/vendor/github.com/thanos-io/objstore/objstore.go b/vendor/github.com/thanos-io/objstore/objstore.go
index 5bcc55f18..8d696aa65 100644
--- a/vendor/github.com/thanos-io/objstore/objstore.go
+++ b/vendor/github.com/thanos-io/objstore/objstore.go
@@ -668,6 +668,9 @@ func newTimingReadCloser(rc io.ReadCloser, op string, dur *prometheus.HistogramV
        // Initialize the metrics with 0.
        dur.WithLabelValues(op)
        failed.WithLabelValues(op)
+       if fetchedBytes != nil {
+               fetchedBytes.WithLabelValues(op)
+       }
        objSize, objSizeErr := TryToGetSize(rc)
        return &timingReadCloser{
                ReadCloser:        rc,
```
Separately there's also now messages about the size not being able to be guessed like:
```
level=warn msg="could not guess file size for multipart upload; upload might be not optimized" name="rules/user-1/dGVzdF9ydWxlX2dyb3VwX25hbWVzcGFjZQ==/dGVzdF9ydWxlX2dyb3Vw" err="unsupported type of io.Reader: io.nopCloserWriterTo"
```
I'll double check my understanding then open a PR to objstore to try to help with the above.

**Comment by andyasp** [comment]
Opened https://github.com/thanos-io/objstore/pull/77 and https://github.com/thanos-io/objstore/pull/78

**Comment by andyasp** [comment]
The above PRs were merged and I revendored. The tests are now passing.

**Review by 56quarters — APPROVED** [review]
LGTM, thank you!

**Pull request merged**